pallets/automation-price/src/lib.rs

```diff
@@ -0,0 +1,700 @@
+				if weight_left < asset_reset_weight {
+					return weight_left
+				}
+				// TODO: this assumes that all assets that need to be reset in a period can all be done successfully in a block.
+				// 			 in the future, we need to make sure to be able to break out of for loop if out of weight and continue 
+				//       in the next block. Right now, we will not run out of weight - we will simply not execute anything if 
+				//       not all of the asset resets can be run at once. this may cause the asset reset triggers to not go off,
+				//       but at least it should not brick the chain.
+				for asset in scheduled_deletion_assets {
+					Self::delete_asset_tasks(asset.clone());
+					Self::update_asset_reset(asset.clone(), current_time_slot);
+					if let Some(last_asset_price) = Self::get_asset_price(asset.clone()) {
+						AssetTargetPrices::<T>::insert(asset.clone(), last_asset_price);
+					};
+					Self::deposit_event(Event::AssetPeriodReset { asset });
+				}
+				ScheduledAssetDeletion::<T>::remove(current_time_slot);
+				weight_left = weight_left - asset_reset_weight;
+			}
+
+			// run as many scheduled tasks as we can
+			let task_queue = Self::get_task_queue();
+			weight_left = weight_left.saturating_sub(T::DbWeight::get().reads(1 as Weight));
+			if task_queue.len() > 0 {
+				let (tasks_left, new_weight_left) = Self::run_tasks(task_queue, weight_left);
+				TaskQueue::<T>::put(tasks_left);
+				weight_left =
+					new_weight_left.saturating_sub(T::DbWeight::get().writes(1 as Weight));
+			}
+			weight_left
+		}
+
+		pub fn update_asset_reset(asset: Vec<u8>, current_time_slot: u64) {
+			if let Some(metadata) = Self::get_asset_metadata(asset.clone()) {
+				let expiration_period: u64 = metadata.expiration_period;
+				// start new duration
+				// 1. schedule new deletion time
+				let new_time_slot = current_time_slot.saturating_add(expiration_period);
+				if let Some(mut future_scheduled_deletion_assets) = Self::get_scheduled_asset_period_reset(new_time_slot) {
+					future_scheduled_deletion_assets.push(asset.clone());
+					<ScheduledAssetDeletion<T>>::insert(new_time_slot, future_scheduled_deletion_assets);
+				} else {
+					let new_asset_list = vec![asset.clone()];
+					<ScheduledAssetDeletion<T>>::insert(new_time_slot, new_asset_list);
+				}
+			};
+		}
+
+		pub fn get_current_time_slot() -> Result<UnixTime, Error<T>> {
+			let now = <timestamp::Pallet<T>>::get().saturated_into::<UnixTime>();
+			if now == 0 {
+				Err(Error::<T>::BlockTimeNotSet)?
+			}
+			let now = now.saturating_div(1000);
+			let diff_to_min = now % 60;
+			Ok(now.saturating_sub(diff_to_min))
+		}
+
+		pub fn delete_asset_tasks(asset: Vec<u8>) -> Weight {
+			// delete scheduled tasks
+			ScheduledTasks::<T>::remove_prefix((asset.clone(), ), None);
+			// delete tasks from tasks table
+			Tasks::<T>::remove_prefix((asset.clone(), ), None);
```

Comment:
this _should_ be 1 write, as per moonbeam. e.g. https://github.com/PureStake/moonbeam/commit/2a5e7e82e01d187ea547c91e51df653f2c7047cb#diff-4568d58ba37a8c54256c2853ef41bdfd28bbd61eb4c5c57b75a8787dd1164d2cR177 (the read is from the if statement above this line)